### PR TITLE
Used Agent to backport PsiModifierList extension to previos buildversions

### PR DESCRIPTION
**Pull request by mplushnikov**
It should be possible to "backport" (inject/patch) the PsiModifierList extension to current and previous builds of IntelliJ.
What do you think?

### CODE REVIEW

**Comment by AlexejK** [comment]
Very interesting idea!
What's the idea behind `LombokPsiModifierListImpl` ? Or was this initial backport attempt before trying the agent approach?

I didn't have a chance to try this yet, but how will the affect testability?

**Comment by mplushnikov** [comment]
Oups, "LombokPsiModifierListImpl" is obsolete. I just forgot to remove it.
I'll investigate issue with failing tests, it's working on my windows mashine, I think it is something with os-specific....

File: src/main/java/de/plushnikov/intellij/plugin/LombokPluginApplicationComponent.java [hunk 1]
```diff
@@ -63,6 +68,25 @@ public void initComponent() {
...
+    final BuildNumber currentBuild = ApplicationInfo.getInstance().getBuild();
+    if(currentBuild.compareTo(BuildNumber.fromString("146.1154")) < 0) {
+      LOG.info("Starting injection of IntelliJ-Patch");
+      injectAgent();
+    }
+  }
+
+  private void injectAgent() {
+    LOG.info("pre injection");
+    String rootOfClass = ClassRootFinder.findClassRootOfClass(IdeaPatcher.class).substring(1);
```
Comment (AlexejK) [inline src/main/java/de/plushnikov/intellij/plugin/LombokPluginApplicationComponent.java]: why are you removing initial slash? This is partialy responsible for test failures on *nix I believe.
What is the output for you on Windows?

File: patcher/src/main/java/de/plushnikov/intellij/plugin/agent/IdeaPatcher.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    System.out.println("Finished IntelliJ Lombok Agent");
+  }
+
+  static void runAgent(String agentArgs, Instrumentation instrumentation, boolean injected) throws Exception {
+
+    IdeaPatcherOptionsHolder optionsHolder = IdeaPatcherOptionsHolder.getInstance();
+    optionsHolder.addAll(agentArgs);
+
+    List<IdeaPatcherTransformer> knownTransformers = new ArrayList<IdeaPatcherTransformer>();
+    knownTransformers.add(new ModifierVisibilityClassFileTransformer());
```
Comment (mplushnikov) [inline patcher/src/main/java/de/plushnikov/intellij/plugin/agent/IdeaPatcher.java]: I think we can remove this collection? It is populated but not used.

File: patcher/src/main/java/de/plushnikov/intellij/plugin/agent/IdeaPatcher.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    for (IdeaPatcherTransformer transformer: KNOWN_TRANSFORMERS) {
+      if (transformer.supported()) {
+        instrumentation.addTransformer(transformer, transformer.canRetransform());
+      }
+    }
+
+    instrumentation.addTransformer(new ModifierVisibilityClassFileTransformer(), true);
```
Comment (mplushnikov) [inline patcher/src/main/java/de/plushnikov/intellij/plugin/agent/IdeaPatcher.java]: Why do we add ModifierVisibilityClassFileTransformer twice? Here and through KNOWN_TRANSFORMERS?

Comment (AlexejK) [inline patcher/src/main/java/de/plushnikov/intellij/plugin/agent/IdeaPatcher.java]: it was a miss. I noticed it myself already and removed in my local. will update soon.

File: patcher/src/main/java/de/plushnikov/intellij/plugin/agent/IdeaPatcher.java (hunk 2 above)
Comment (AlexejK) [inline patcher/src/main/java/de/plushnikov/intellij/plugin/agent/IdeaPatcher.java]: yepp, old code, a bit dirty

**Comment by AlexejK** [comment]
Im looking into why IDE 14 is failing, will update you soon =)

**Comment by AlexejK** [comment]
ValModifierTest is the only one failing on val type inspection.
2 options here: 
- stop expecting the type in the modifier test
- figure out what changed between 14 and 15

Problem is:
In 14 it resolves to: `com.intellij.psi.impl.source.PsiClassReferenceType`
In 15 it resolves to: `com.intellij.psi.PsiPrimitiveType`

Suggestions?

**Comment by mplushnikov** [comment]
Yes, I was look into ValModifierTest also.
So far as I can see there is a problem with internal caching of Intellij. At first var.getType() evaluation the type is unknown. At second evaluation, after typing, it doesn't go inside of ValProcessor.

**Comment by mplushnikov** [comment]
I think it was fixed in IntelliJ by https://youtrack.jetbrains.com/oauth?state=%2Fissue%2FIDEA-147214

**Comment by AlexejK** [comment]
Yes, I see that between 14.1 and 15.0 the change was moving to the CacheValueProvider:
https://github.com/JetBrains/intellij-community/commit/6569cee283ceb86ebdd3df2a65df2a72582364ee

I was thinking if we should "rewrite" that class method. But it requires lower level javassist api instead :( as it needs Generics.

Another way is having a "backport" method in plugin code-base and delegate to it instead.

**Comment by Janmm14** [comment]
youcould look into aving the rewritten class compiled under resources and then replace the whole class or sth like that.

**Comment by AlexejK** [comment]
It is a way around it (I was considering that), But the question is if we will cause more issues by doing that. A few other changes have been made to that class. But I guess by utilizing `@SupportedBuild(max = "143.708")`  it might be safe enough.

**Comment by mplushnikov** [comment]
I also think we should not change val processing. It's available in Intellij 15 and in the current version, that's fine enough. If backported PsiModifierList will work for Value and other "final" annotations, it will be already a huge improvement.

**Comment by AlexejK** [comment]
It does work according to my tests. Let me fix up the test then and update

**Comment by AlexejK** [comment]
FYI: There is one minor inconvenience with the agent approach when developing the plugin itself on versions that require backport support. Gradle build is able to run with help of agent, but the way IntelliJ runs tests inside of the UI, it points to classes directory instead of jar dependency and thus causes agent not to load.

Considering agent is used as a "workaround", I think it should be acceptable that only gradle-initiated tests will go through at this point. Might be able to fiddle around with `idea.module {}` in `build.gradle` to get this working but not sure if it's worth the effort.

**Comment by AlexejK** [comment]
To be honest, for the Reflection I think it's somewhat fine for the backport.
We are running in a different context and it might be a bit tricky to do that via javassist.
We could ofc explore changing class files via cglib instead.

But this should be good-enough to merge?
Or do you have some additional things you want to change?

**Comment by AlexejK** [comment]
Shall we go ahead and merge this and fix possible issues on top of master?

**Comment by mplushnikov** [comment]
Yes, I just merged it.

**Pull request merged**